modules/querier/querier.go

```diff
@@ -271,30 +282,45 @@ func (q *Querier) FindTraceByID(ctx context.Context, req *tempopb.TraceByIDReque
 }
 
 // forGivenIngesters runs f, in parallel, for given ingesters
-func (q *Querier) forGivenIngesters(ctx context.Context, replicationSet ring.ReplicationSet, f func(client tempopb.QuerierClient) (interface{}, error)) ([]responseFromIngesters, error) {
-	results, err := replicationSet.Do(ctx, q.cfg.ExtraQueryDelay, func(ctx context.Context, ingester *ring.InstanceDesc) (interface{}, error) {
+func (q *Querier) forGivenIngesters(ctx context.Context, replicationSet ring.ReplicationSet, f func(ctx context.Context, client tempopb.QuerierClient) (interface{}, error)) ([]responseFromIngesters, error) {
+	if ctx.Err() != nil {
+		_ = level.Debug(log.Logger).Log("foreGivenIngesters context error", "ctx.Err()", ctx.Err().Error())
+		return nil, ctx.Err()
+	}
+
+	span, ctx := opentracing.StartSpanFromContext(ctx, "Querier.forGivenIngesters")
+	defer span.Finish()
+
+	doFunc := func(funcCtx context.Context, ingester *ring.InstanceDesc) (interface{}, error) {
+		if funcCtx.Err() != nil {
+			_ = level.Warn(log.Logger).Log("funcCtx.Err()", ctx.Err().Error())
+			return nil, funcCtx.Err()
+		}
+
 		client, err := q.pool.GetClientFor(ingester.Addr)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, fmt.Sprintf("failed to get client for %s", ingester.Addr))
 		}
 
-		resp, err := f(client.(tempopb.QuerierClient))
+		resp, err := f(funcCtx, client.(tempopb.QuerierClient))
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrap(err, fmt.Sprintf("failed to execute f() for %s", ingester.Addr))
 		}
 
 		return responseFromIngesters{ingester.Addr, resp}, nil
-	})
+	}
+
+	results, err := replicationSet.Do(ctx, q.cfg.ExtraQueryDelay, doFunc)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "failed to get response from ingesters")
 	}
 
 	responses := make([]responseFromIngesters, 0, len(results))
 	for _, result := range results {
 		responses = append(responses, result.(responseFromIngesters))
 	}
 
-	return responses, err
+	return responses, nil
```

Comment:
Double check me here, but I think we already return all the possible `err` values.